### DISCUSSION TITLE
"At Least Once" or "Exactly Once"?

### DISCUSSION THREAD

**Discussion by v13leo** [jobrunr/jobrunr#358] [Q&A] [answered]
Hey, I wanted to know what the processing semantics are for JobRunr and I seem to have found some conflicts and wanted to get clarity on it.

This screenshot from the homepage seems to say that the jobs will run "At Least Once", which indicates to me that it's possible for a single job to be processed in duplicate or more.
<img width="409" alt="Screen Shot 2022-02-09 at 11 39 10 AM" src="https://user-images.githubusercontent.com/11339485/153248521-042992b0-aae9-4d47-b6a0-a76f7d64376f.png">

Deeper in the documentation it says that it will only run "Exactly Once"
<img width="874" alt="Screen Shot 2022-02-09 at 11 37 10 AM" src="https://user-images.githubusercontent.com/11339485/153248688-e93fdf02-48b7-45fd-bc08-418dc44e79dd.png">

Then your reddit comment seems to say that the default behavior is "At Least Once"
https://www.reddit.com/r/java/comments/gs4l50/jobrunr_available_in_maven_central/fs3af0k/
Is this accurate? Is there a way to change this default behavior to "Exactly Once"?

Then your blog post from last year says that it's "Exactly Once"
https://blogs.oracle.com/javamagazine/post/introducing-jobrunr-a-distributed-job-scheduler-for-java
`Since the BackgroundJobServers use optimistic locking, each job will be processed only once` 
...which seems to indicate that jobs processing won't be duplicated.

Can I get some clarity on this? I'm assuming that by using persistent db jobs with optimistic locking that each job can only be run once without the risk of two executors picking up the same job and running the job twice, but I wanted to make sure.

**Comment by rdehuyss** [comment, marked as answer]
Hi @Camo13ammo , thanks for bringing that up to my attention. It's definitely something that I can improve on the website.

If no exceptions during the run of your job, JobRunr will process your job exactly once by means of optimistic locking. 
If however your job is existing out of multiple phases and one of those last phases fails, all the prior phases will be re-executed when your job is retried.

So, to answer your question: exactly once when all goes well, at least once if there is an exception.

**Reply by v13leo** [reply]
Thank you for your response!

When you say multiple phases, do you mean something like...
```
jobScheduler.enqueue(() -> {
    this.executeFirstPhase();
    this.executeSecondPhase();
});
```
If `this.executeSecondPhase();` throws an exception, `this.executeFirstPhase();` will, of course be executed again when the entire job is reprocessed.

Or does "multiple phases" mean when you are Job Chaining that the first job will get re-run? 
```
BackgroundJob
        .enqueue(() -> archiveService.createArchive(folder))
        .continueWith(() -> notifyService.notifyViaSlack("ops-team", "The following folder was archived: " + folder))
```
It seems that based on how job chaining works with AWAIT status, that if the `notifyViaSlack` job fails, it won't then cause the `createArchive` job to run again.

Or does "multiple phases" mean something else?

**Reply by rdehuyss** [reply]
I'm talking about your first example, i.e. 

```java
jobScheduler.enqueue(() -> {
    this.executeFirstPhase();
    this.executeSecondPhase();
});
```

Regarding the job chaining: no - as these are also jobs, they follow the same principle with optimistic locking.

I never thought that the "multiple phases" would be open for so many interpretations 😄 . Sorry for that confusion and good that you double checked!

**Reply by v13leo** [reply]
Thank you! I appreciate the help!

**Reply by azeem87** [reply]
so it means failure of notifyService.notifyViaSlack will not cause archiveService.createArchive job to rerun in multiple phase

**Reply by rdehuyss** [reply]
@azeem87 - indeed, the `archiveService.createArchive` is an atomic job and so is the `notifyService.notifyViaSlack` - if the `notifyService.notifyViaSlack` it will just retry only that job until it succeeds (or exhausts the retries).
